### PR TITLE
Fix trade record

### PR DESCRIPTION
**Pull request by rwcarlsen**
Previously, if a facility modified a received resource in its Accept[Matl/Prod]Trades function, DRE wouldn't record the traded resource until after AcceptTrades returned - resulting in the incorrect resource id and state being recorded as what was transacted.  This tweaks DRE to record the trades and resource info after trades are completely resolved but before the materials are sent to the requesters.  Very tricky bug to find!

### CODE REVIEW

**Comment by gidden** [comment]
Thanks for the leg work, @rwcarlsen. Could you slap a deprecated warning in the documentation of the old API?

**Comment by rwcarlsen** [comment]
done.

**Comment by gidden** [comment]
Testing this would either require a full integration test (effectively what is in cyclus/cycamore#333) or a mock resource exchange (which we don't have, but I opened #1147) combined with a test context and mock back end. I'm okay with accepting the two PRs in conjunction as needed.

**Comment by rwcarlsen** [comment]
Right.  I think the cycamore fab test will have to be good enough for now.
This is an important bug fix, and we unfortunately don't have the time now
for a better test.

**Comment by gidden** [comment]
Hmm, I'm getting an ABI test error with
```
-cyclus::TradeExecutor<cyclus::Material>::ExecuteTrades()
-cyclus::TradeExecutor<cyclus::Product>::ExecuteTrades()
```

**Comment by gidden** [comment]
So, this fails for me locally because the old `ExecuteTrades()` never is seen in a execution pathway, therefore may not be getting picked up as part of the symbols. If I add a fake pathway, e.g.,
```
    // execute trades!                                                                                                                                                                   
    TradeExecutor<T> exec(trades);
    if (trades.size() < 0) {
      exec.ExecuteTrades();
      exec.RecordTrades(ctx_);
    }
    exec.ExecuteTrades(ctx_);
```
then all tests pass.

**Comment by gidden** [comment]
I can't think of anything that isn't a completely hacky work around. Can you just add both of those functions to the `api_blacklist` in `symbchk.py`?

**Comment by rwcarlsen** [comment]
blacklisted.

File: src/trade_executor.h [hunk 1]
```diff
@@ -51,12 +51,25 @@ class TradeExecutor {
 
   /// @brief execute all trades, collecting responders from bidders and sending
   /// responses to requesters
+  /// @deprecated Use ExecuteTrades(Context*) instead.
   void ExecuteTrades() {
+    Warn<IO_WARNING>("this function does not record trades to the database");
     GroupTradesBySupplier(trade_ctx_, trades_);
```
Comment (gidden) [inline src/trade_executor.h]: @rwcarlsen, it looks like this function is just implementing what is now `ExecuteTrades(NULL)`. Can you update this to reflect that?

File: src/trade_executor.h [hunk 2]
```diff
@@ -51,12 +51,25 @@ class TradeExecutor {
 
   /// @brief execute all trades, collecting responders from bidders and sending
   /// responses to requesters
+  /// @deprecated Use ExecuteTrades(Context*) instead.
   void ExecuteTrades() {
+    Warn<IO_WARNING>("this function does not record trades to the database");
```
Comment (gidden) [inline src/trade_executor.h]: @rwcarlsen, it looks like this function is just implementing what is now ExecuteTrades(NULL). Can you update this to reflect that?

**Comment by gidden** [comment]
ok, this passes now. final comment then ready to merge.

File: src/trade_executor.h (hunk 2 above)
Comment (rwcarlsen) [inline src/trade_executor.h]: done.

**Comment by gidden** [comment]
Ah sorry, I should have been clearer. I meant that the function seemed to be doing exactly what `ExecuteTrades(NULL)` would do. I think one of two things should happen (you choose):

- remove body of `ExecuteTrades()`
- update body of `ExecuteTrades()` to invoke `ExecuteTrades(NULL)`

**Comment by rwcarlsen** [comment]
done.

**Comment by gidden** [comment]
thanks @rwcarlsen

**Pull request merged**